frontend/src/modules/entitieslist/components/EntitiesList.tsx

```diff
@@ -45,91 +47,79 @@ type InternalProps = Props & {
  * entities. It interacts with `core/navigation` when an entity is selected.
  *
  */
-export class EntitiesListBase extends React.Component<InternalProps> {
-    list: { current: any };
-
-    constructor(props: InternalProps) {
-        super(props);
-        this.list = React.createRef();
-    }
-
-    componentDidMount() {
-        document.addEventListener('keydown', this.handleShortcuts);
-
-        this.selectFirstEntityIfNoneSelected();
-    }
-
-    componentWillUnmount() {
-        document.removeEventListener('keydown', this.handleShortcuts);
-    }
-
-    componentDidUpdate(prevProps: InternalProps) {
-        this.selectFirstEntityIfNoneSelected();
-
-        // Whenever the route changes, we want to verify that the user didn't
-        // change locale, project, resource... If they did, then we'll have
-        // to reset the current list of entities, in order to start a fresh
-        // list and hide the previous entities.
-        //
-        // Notes:
-        //  * It might seem to be an anti-pattern to change the state after the
-        //    component has rendered, but that's actually the easiest way to
-        //    implement that feature. Note that the first render is not shown
-        //    to the user, so there should be no blinking here.
-        //    Cf. https://reactjs.org/docs/react-component.html#componentdidupdate
-        //  * Other solutions might involve using `connected-react-router`'s
-        //    redux actions (see https://stackoverflow.com/a/37911318/1462501)
-        //    or using `history.listen` to trigger an action on each location
-        //    change.
-        //  * I haven't been able to figure out how to test this feature. It
-        //    is possible that going for another possible solutions will make
-        //    testing easier, which would be very desirable.
-        const previous = prevProps.parameters;
-        const current = this.props.parameters;
-        if (
-            previous.locale !== current.locale ||
-            previous.project !== current.project ||
-            previous.resource !== current.resource ||
-            previous.search !== current.search ||
-            previous.status !== current.status ||
-            previous.extra !== current.extra ||
-            previous.tag !== current.tag ||
-            previous.author !== current.author ||
-            previous.time !== current.time
-        ) {
-            this.props.dispatch(entities.actions.reset());
-        }
-
-        // Scroll to selected entity when entity changes
-        // and when entity list loads for the first time
-        if (
-            previous.entity !== current.entity ||
-            (!prevProps.entities.entities.length &&
-                this.props.entities.entities.length)
```

Comment:
In this case we need to actually compare the value of `entities.entities.length` between its previous value and its current value, so only watching for a change of any kind will not get the results needed.

If you utilize the `usePrevious` hook (see example of implementation [here](https://github.com/mozilla/pontoon/blob/master/frontend/src/modules/machinery/components/Machinery.tsx#L55)) to store the previous value of `props.entities.entities`, check for changes in `props.entities.entities` in the dependency array, and then wrap the below code in an `if` statement that makes the same checks as above the test errors around the mediaQuery are no longer present. 

There are still failing tests but those appear to be related to receiving improper values (which are likely legitimate fails that are occurring due to the changes made)

Note: I tested this off of commit [ed684ff](https://github.com/mozilla/pontoon/pull/2413/commits/ed684ffabe147686250d7bfa482c270db764d966)

Reply:
Thanks for the feedback @abowler2 but i have resolved that already, i am only having issues testing the arguments contained within a dispatched action from the method getMoreEntities() and toggleBatchEdititng(). I can't seem to access the actions which is valid because, it's a functional component so i can't get an instance of it, functions within is not exposed to other files. I am basically trying to workaround this and test it in other way if possible. Any suggestions from you will be welcomed.

PS: I am not good with tests, i am currently learning so i could get around this.

Reply:
My apologies. I do see that you made that change in a later commit (when I first ran the test suite locally I was failing around the media query, so I opted to start at the commit before your test changes. Clearly my build was off and did not pick up all of your changes as running it from the latest commit now is fine)

That said, it looks like you are on the right track in that you seem to have noted that it is the use of Enzyme's `instance()` method that is failing since that [returns null](https://enzymejs.github.io/enzyme/docs/api/ShallowWrapper/instance.html#instance--reactcomponent) in functional components.

Sadly, I don't have enough testing experience either to be able to steer you towards a clear path.  I suspect that we may want to think about approaching the test differently rather than shoehorning the older method with a work around.

Maybe @eemeli or @mathjazz could provide some guidance here?

Reply:
@abowler2 Exactly, that was the problem i was facing.

Reply:
In order to get access to the "instance" of a functional component, [React.useImperativeHandle](https://reactjs.org/docs/hooks-reference.html#useimperativehandle) is your friend.

The code around this should include a comment noting that this exposure is solely for testing, and really the upstream component `react-infinite-scroller` ought to play more nicely with test environment, so that the test could be made more real.

Reply:
@eemeli I did that as well using forwardRefs and imperativehandle if you check the comments above, already stated that... But the issue now lies in getting the actions dispatched within those functions exposed.. When I did that, it's sayings actions not defined... Maybe you can help check it out.. Nonetheless I have skipped this test by wrapping setTimeOut() around it,that is why I could pass all test cases till I figure that out... My take is that, we should find another way of testing it rather than trying to access the arguments dispatched by an action.

Reply:
> really the upstream component react-infinite-scroller ought to play more nicely with test environment, so that the test could be made more real.

@eemeli Given that the underlying reason we are converting this component is to replace this infinite scroll component (which requires a work around to properly load new content and is no longer maintained well) with the [hook version](https://github.com/mozilla/pontoon/pull/2394) that is now used in Concordance search (Machinery.tsx) should we consider just replacing it within this work?  

With Machinery we had a button to load more content so the conversion could be decoupled from the implementation of the infinite scroll hook.

Reply:
As @Pike [pointed out](https://github.com/mozilla/pontoon/pull/2413#issuecomment-1025546448), this patch is not even required for using `react-infinite-scroll-hook`, because `EntitiesList` actually already is a functional component. Which I overlooked, sorry about that. :/

I'd still proceed (because hooks seem to be better supported by the React ecosystem), unless it will make transition to `react-infinite-scroll-hook` harder. @abowler2 Thoughts?

We should not worry about the skipped `excludes current entities when requesting new entities` test, because it's only required by `InfiniteScroll`, which we'll drop. I'm fine with dropping that test. I'm also fine with doing the switch within the same patch if that makes things easier.

Reply:
> As @Pike [pointed out](https://github.com/mozilla/pontoon/pull/2413#issuecomment-1025546448), this patch is not even required for using `react-infinite-scroll-hook`, because `EntitiesList` actually already is a functional component. Which I overlooked, sorry about that. :/

`EntitiesList` is a functional component; however, the infinite scroll component exists in `EntitiesListBase` which is a class component. Unless I'm missing something, `EntitiesListBase` would need to be functional in order to utilize the hook

> I'd still proceed (because hooks seem to be better supported by the React ecosystem), unless it will make transition to `react-infinite-scroll-hook` harder. @abowler2 Thoughts?

Given that current and future iterations of React are hook based I would agree that the effort to convert this component is worth it.

> We should not worry about the skipped `excludes current entities when requesting new entities` test, because it's only required by `InfiniteScroll`, which we'll drop. I'm fine with dropping that test. I'm also fine with doing the switch within the same patch if that makes things easier.

If that test is only a result of the `InfiniteScroll` component, then we should definitely just make the switch now rather than spend any more time trying to 'fix' a test that will soon be obsolete or require a new approach.

Reply:
AFAICT, all of

https://github.com/mozilla/pontoon/blob/f867702b8f917fbb4ceda5fbd16996a2a6783d2e/frontend/src/modules/entitieslist/components/EntitiesList.tsx#L406-L420

is just unhooking hooks, and then passing them as props to the class component. We'd do the same thing for `react-infinite-scroll`

Reply:
Correct. We could put the `react-infinite-scroll`-related logic into `EntitiesList`, including the `getMoreEntities` function, and keep the majority of code in `EntitiesListBase` as a class component. In the position we're in right now with this patch, I'd do that as a last resort, though.

> If that test is only a result of the `InfiniteScroll` component, then we should definitely just make the switch now rather than spend any more time trying to 'fix' a test that will soon be obsolete or require a new approach.

It is. We'll need to change the API call on backend to only return the requested page (instead of exluding already received entities).